en_US/advanced/lang-exproto.md

```diff
@@ -0,0 +1,144 @@
+    - Provides the `StartTimer` interface. Called by user's server to start a timer such as a heartbeat for the connected process.
+    - Provides the `Publish` interface. Called by user's server to publish messages to the EMQ X.
+    - Provides the `Subscribe` interface. Called by user's server to subscribe to a topic to receive certain downlink messages from the EMQ X.
+    - Provides the `Unsubscribe` interface. Called by user's server to unsubscribe from a topic.
+    - Calls the `OnTimerTimeout` callback. Used to handle timer timeout events
+    - Call the `OnReceivedMessages` callback. Used to receive downlink messages.
+
+
+## APIs
+
+From a gRPC perspective, ExProto acts as a client to send callback requests to the `ConnectionHandler` service. It also acts as a server side to provide the `ConnectionAdapter` service to user's server to provide calls to various interfaces. As shown in the figure.
+
+![Extension Protocol gRPC Arch](../modules/assets/exproto-grpc-arch.jpg)
+
+The services defination see: [exproto.proto](https://github.com/emqx/emqx/blob/v4.3-beta.1/apps/emqx_exproto/priv/protos/exproto.proto)
+
+For examples:
+
+```protobuff
+syntax = "proto3";
+
+package emqx.exproto.v1;
+
+// The Broker side serivce. It provides a set of APIs to
+// handle a protcol access
+service ConnectionAdapter {
+
+  // -- socket layer
+
+  rpc Send(SendBytesRequest) returns (CodeResponse) {};
+
+  rpc Close(CloseSocketRequest) returns (CodeResponse) {};
+
+  // -- protocol layer
+
+  rpc Authenticate(AuthenticateRequest) returns (CodeResponse) {};
+
+  rpc StartTimer(TimerRequest) returns (CodeResponse) {};
+
+  // -- pub/sub layer
+
+  rpc Publish(PublishRequest) returns (CodeResponse) {};
+
+  rpc Subscribe(SubscribeRequest) returns (CodeResponse) {};
+
+  rpc Unsubscribe(UnsubscribeRequest) returns (CodeResponse) {};
+}
+
+service ConnectionHandler {
+
+  // -- socket layer
+
+  rpc OnSocketCreated(SocketCreatedRequest) returns (EmptySuccess) {};
+
+  rpc OnSocketClosed(SocketClosedRequest) returns (EmptySuccess) {};
+
+  rpc OnReceivedBytes(ReceivedBytesRequest) returns (EmptySuccess) {};
+
+  // -- pub/sub layer
+
+  rpc OnTimerTimeout(TimerTimeoutRequest) returns (EmptySuccess) {};
+
+  rpc OnReceivedMessages(ReceivedMessagesRequest) returns (EmptySuccess) {};
+}
+```
+
+
+## Developing Guide
+
+The user needs to implement the gRPC service of `ConnectionHandler` to receive callback events from EMQ X.
+
+The main development steps are as following:
+
+1. Copy the `lib/emqx_exproto-<x.y.z>/priv/protos/exproto.proto` file to your project.
+2. Generate the code for the gRPC server and client side of `exproto.proto` using the gRPC framework for the corresponding programming language.
+3. Implement the interfaces defined in exhook.proto on demand
+
+Once the development is complete, the service needs to be deployed to a server that can communicate with EMQ X.
```

Comment:
that can communicate with EMQ X. -> that can communicate with EMQ X and ensure that the ports are open.